src/GraphQL.NewtonsoftJson/ExecutionResultJsonConverter.cs

```diff
@@ -12,34 +12,95 @@ public class ExecutionResultJsonConverter : JsonConverter
         {
             var result = (ExecutionResult)value;
 
             writer.WriteStartObject();
 
-            WriteErrors(result.Errors, writer, serializer);
-            WriteData(result, writer, serializer);
-            WriteExtensions(result, writer, serializer);
+            WriteErrors(writer, result.Errors, serializer);
+            WriteData(writer, result, serializer);
+            WriteExtensions(writer, result, serializer);
 
             writer.WriteEndObject();
         }
 
-        private void WriteData(ExecutionResult result, JsonWriter writer, JsonSerializer serializer)
+        private void WriteData(JsonWriter writer, ExecutionResult result, JsonSerializer serializer)
         {
             if (result.Executed)
             {
                 writer.WritePropertyName("data");
-                serializer.Serialize(writer, result.Data);
+                if (result.Data is ExecutionNode executionNode)
+                {
+                    WriteExecutionNode(writer, executionNode, serializer);
+                }
+                else
+                {
+                    serializer.Serialize(writer, result.Data);
+                }
+            }
+        }
+
+        private void WriteExecutionNode(JsonWriter writer, ExecutionNode node, JsonSerializer serializer)
+        {
+            if (node is ValueExecutionNode valueExecutionNode)
+            {
+                serializer.Serialize(writer, valueExecutionNode.ToValue());
+            }
+            else if (node is ObjectExecutionNode objectExecutionNode)
+            {
+                if (objectExecutionNode.SubFields == null)
+                {
+                    writer.WriteNull();
+                }
+                else
+                {
+                    writer.WriteStartObject();
+                    foreach (var childNode in objectExecutionNode.SubFields)
+                    {
+                        writer.WritePropertyName(childNode.Name);
+                        WriteExecutionNode(writer, childNode, serializer);
+                    }
+                    writer.WriteEndObject();
+                }
+            }
+            else if (node is ArrayExecutionNode arrayExecutionNode)
+            {
+                var items = arrayExecutionNode.Items;
+                if (items == null)
+                {
+                    writer.WriteNull();
+                }
+                else
+                {
+                    writer.WriteStartArray();
+                    foreach (var childNode in items)
+                    {
+                        WriteExecutionNode(writer, childNode, serializer);
+                    }
+                    writer.WriteEndArray();
+                }
+            }
+            else if (node == null)
+            {
+                writer.WriteNull();
+            }
+            else
+            {
+                serializer.Serialize(writer, node.ToValue());
```

Comment:
That is, in fact, this line will never be executed. In general, the whole benefit is to call `ToValue` as rarely as possible. Now it only happens for `ValueExecutionNode`. Right?

Reply:
It should not, unless another execution strategy sets Data to something other than a RootExecutionNode. There are probably various tests that still use this code, but I didn’t check other than to change the tests I saw and ensure that both serializers were being exercised. 

Yes, that is correct that ToValue is only called for ValueExecutionNode instance (or a derived instance that does not derive from ObjectExecutionNode or ArrayExecutionNode).